### PR TITLE
Mypy for meadowflow

### PR DESCRIPTION
**Pull request by kurtschelfthout**

### CODE REVIEW

File: src/meadowflow/server/client.py [hunk 1]
```diff
@@ -1,7 +1,10 @@
+from __future__ import annotations
+
 import asyncio
 import pickle
 import time
-from typing import List, Optional
+from types import TracebackType
+from typing import Any, Coroutine, List, Literal, Optional, Type, cast
```
Comment (hrichardlee) [inline src/meadowflow/server/client.py]: Any, Coroutine not used

File: src/meadowflow/server/server_main.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import multiprocessing
 import asyncio
 import contextlib
-from typing import ContextManager
+from typing import ContextManager, Iterator, Optional
```
Comment (hrichardlee) [inline src/meadowflow/server/server_main.py]: ContextManager no longer used

File: src/meadowflow/jobs.py [hunk 3]
```diff
@@ -10,14 +10,17 @@
 from typing import (
     Final,
     Iterable,
+    Protocol,
```
Comment (hrichardlee) [inline src/meadowflow/jobs.py]: Not used

File: src/meadowflow/jobs.py [hunk 4]
```diff
@@ -10,14 +10,17 @@
...
     Sequence,
     List,
     Any,
     Dict,
     Optional,
+    TypedDict,
```
Comment (hrichardlee) [inline src/meadowflow/jobs.py]: Not used

File: src/meadowflow/jobs.py [hunk 5]
```diff
@@ -371,9 +375,10 @@ async def execute(
             # TODO meadowflow needs to store credentials the same way that the
             #  meadowgrid coordinator does, or share them, or something. Currently
             #  we are passing None because we don't have any.
-            credentials_dict = {}
+            credentials_dict: CredentialsDict = {}
 
             # convert a job_function into a job_runner_function
+            job_runner_function: Union[LocalFunction, MeadowGridDeployedRunnable]
```
Comment (hrichardlee) [inline src/meadowflow/jobs.py]: Probably better to call this a `JobRunnerFunction`?

File: src/meadowflow/scheduler.py [hunk 6]
```diff
@@ -1,22 +1,28 @@
...
 import asyncio
 import dataclasses
 import itertools
 import threading
 import traceback
 from asyncio import Task
+from types import TracebackType
 from typing import (
     Dict,
+    Final,
```
Comment (hrichardlee) [inline src/meadowflow/scheduler.py]: Not used

File: src/meadowflow/scheduler.py [hunk 7]
```diff
@@ -1,22 +1,28 @@
...
     List,
     Tuple,
     Iterable,
     Optional,
     Callable,
     Awaitable,
     Sequence,
```
Comment (hrichardlee) [inline src/meadowflow/scheduler.py]: Not used

File: src/meadowflow/scheduler.py [hunk 8]
```diff
@@ -1,22 +1,28 @@
...
+    Type,
     Union,
     Literal,
     Any,
```
Comment (hrichardlee) [inline src/meadowflow/scheduler.py]: Not used

File: src/meadowflow/time_event_publisher.py [hunk 9]
```diff
@@ -510,18 +516,25 @@ def apply_to_date(d: datetime.date) -> datetime.datetime:
...
                             time_of_day.time_zone,
                             date,
                             occurrence,
                         ),
-                    ),
+                    )
+
+                self._call_at.call_at(
+                    occurrence,
+                    # capture the value of date and occurrence
```
Comment (hrichardlee) [inline src/meadowflow/time_event_publisher.py]: Comment should be moved up into `append`

File: src/meadowflow/topic.py [hunk 10]
```diff
@@ -4,7 +4,17 @@
...
 from dataclasses import dataclass
-from typing import Iterable, Sequence, List, Callable, Any
+from typing import (
+    Generic,
+    Iterable,
+    Optional,
+    Sequence,
+    List,
+    Callable,
+    Any,
```
Comment (hrichardlee) [inline src/meadowflow/topic.py]: Not used

File: tests/test_meadowflow/test_deployment_override.py [hunk 11]
```diff
@@ -1,6 +1,7 @@
-from typing import Union
+from typing import List, Union, cast
```
Comment (hrichardlee) [inline tests/test_meadowflow/test_deployment_override.py]: List not used

File: tests/test_meadowflow/test_deployment_override.py [hunk 12]
```diff
@@ -1,6 +1,7 @@
...
+from meadowflow.event_log import Event
```
Comment (hrichardlee) [inline tests/test_meadowflow/test_deployment_override.py]: Not used

File: tests/test_meadowflow/test_deployment_override.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
...
 
-from meadowflow.jobs import Job, JobRunOverrides
+from meadowflow.jobs import Job, JobPayload, JobRunOverrides
```
Comment (hrichardlee) [inline tests/test_meadowflow/test_deployment_override.py]: JobPayload not used

File: tests/test_meadowflow/test_scheduler.py [hunk 14]
```diff
@@ -599,7 +600,7 @@ def _run_instantiate_date_scope() -> ScopeValues:
 
 
 @add_scope_jobs_decorator
-def _run_add_scope_jobs(scope: ScopeValues) -> Sequence[Job]:
+def _run_add_scope_jobs(scope: ScopeValues, /) -> Sequence[Job]:
```
Comment (hrichardlee) [inline tests/test_meadowflow/test_scheduler.py]: Is this needed? I think I was able to take it out without causing any problems

**Review by hrichardlee — APPROVED** [review]
As usual my contribution is pointing out imports that aren't used :)

**Comment by kurtschelfthout** [comment]
That's actually a soul-crushingly boring thing to have to point out, I'll have to figure something out to organize those imports.

File: src/meadowflow/jobs.py (hunk 5 above)
Comment (kurtschelfthout) [reply src/meadowflow/jobs.py]: didn't get this one, this is a variable not a type?

Comment (hrichardlee) [reply src/meadowflow/jobs.py]: I think `Union[LocalFunction, MeadowGridDeployedRunnable] == JobRunnerFunction`

File: tests/test_meadowflow/test_scheduler.py (hunk 14 above)
Comment (kurtschelfthout) [reply tests/test_meadowflow/test_scheduler.py]: No - I had a hard time typing that decorator so was trying all sorts of stuff.

File: src/meadowflow/jobs.py (hunk 5 above)
Comment (kurtschelfthout) [reply src/meadowflow/jobs.py]: oh, right, good catch

**Pull request merged**